### PR TITLE
CWAP-215:  Introducing basic document for storing externally sourced …

### PR DESCRIPTION
**Pull request by dkichler**
…data as "shoebox snippets"

### CODE REVIEW

File: databases/business-sync/fragment-shoebox-snippet.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      required: true,
+      mustNotBeEmpty: true
+    },
+    source: {
+      // The source of the snippet
+      type: 'string',
+      required: true,
+      mustNotBeEmpty: true
+    },
+    kashooEntityType: {
```
Comment (dkichler) [inline databases/business-sync/fragment-shoebox-snippet.js]: I almost wondered if this should be an array of objects, each with a type and ID

File: databases/business-sync/fragment-shoebox-snippet.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+{
+  channels: getDocSyncChannels(doc, oldDoc),
+  typeFilter: function(doc, oldDoc) {
+    return createBusinessEntityRegex('shoeboxSnippet\\.[A-Za-z0-9_-]+\\.[A-Za-z0-9_-]+').test(doc._id);
```
Comment (dkichler) [inline databases/business-sync/fragment-shoebox-snippet.js]: keyspace schema:  `biz.<biz_id>.shoeboxSnippet.<snippet_type>.<uuid>`

File: databases/business-sync/doc-definitions.js [hunk 3]
```diff
@@ -33,17 +33,22 @@ function() {
...
-      };
+      if (basePrivilegeName) {
+        return {
+          view: [ toSyncChannel(businessId, 'VIEW_' + basePrivilegeName), staffChannel ],
+          add: [ toSyncChannel(businessId, 'ADD_' + basePrivilegeName), staffChannel ],
+          replace: [ toSyncChannel(businessId, 'CHANGE_' + basePrivilegeName), staffChannel ],
+          remove: [ toSyncChannel(businessId, 'REMOVE_' + basePrivilegeName), staffChannel ]
+        };
+      } else {
+        return { write: staffChannel };
```
Comment (ziemek) [inline databases/business-sync/doc-definitions.js]: Just want to confirm: write privilege automatically also grants read privilege to the staff channel? (and only the staff channel)

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 2 above)
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: Can we add this as a comment. Most importantly to explain the two regex fields.

Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: I feel that `shoeboxSnippet` is a bit wordy and maybe anachronistic (i like using that word). Is it your preferred noun? Should we brainstorm for other nouns? `dataSnippet`, `rawData`, `sourceData` ...? :brain: ⛈

File: databases/business-sync/fragment-shoebox-snippet.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  },
+  allowUnknownProperties: true,
+  immutable: true,
+  propertyValidators: {
+    type: {
+      // The type of shoebox snippet (ie, bank transaction, email, photo, digital receipt)
+      type: 'string',
+      required: true,
+      mustNotBeEmpty: true
+    },
```
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: Thoughts about making this an enum for extra data safety? Would it be easy to extend if it were an enum?

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 1 above)
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: Nice idea on having the entity type and id 👍  We can ask chuck and amrik whehter it might be possible to reference two entities. I can't think of a reason. It seems that everything will just be associated with a record (though not always, for example a tax notice of assessment).

File: databases/business-sync/fragment-shoebox-snippet.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      type: 'integer',
+      required: false,
+      minimumValue: 1
+    },
+    data: {
+      // Raw data of the snippet
+      type: 'string',
+      required: true,
+      mustNotBeEmpty: true
+    }
```
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: How about another field for data updates? It seems that data could be the most up to date version and we could have a history array?

Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: How about a date field that's a timestamp for when this data came in?

Also how about a processed flag? That's not necessarily implied by kashooId.

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 1 above)
Comment (dkichler) [inline databases/business-sync/fragment-shoebox-snippet.js]: I'm now questioning whether we want to even have this association stored here, and that this document really just stores the raw data as received from the external source.  It makes more sense to me that a Kashoo entity only gets associated with raw data when we decide to do something with it, which will likely include a bunch of additional metadata, and some of which you've alluded to in comments below (processed timestamp, etc).  Upon that operation happening, we should probably store a new document including the associations and relevant metadata.  My thinking is to keep document types as specific and simple as possible, with only a single purpose (in this case, simply a record of when/what external data was received)

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 2 above)
Comment (dkichler) [inline databases/business-sync/fragment-shoebox-snippet.js]: Comment to come on regex.

On the topic of `shoeboxSnippet`, I went with that because, as silly as it might sound, it ties in very closely with a concept/analogy that has been a motivator behind Kashoo pretty much since the beginning, and I think also ties in well with the 'smart inbox' idea.  I'm not overly attached to it, so if you feel strongly about it I'm open to suggestions.

File: databases/business-sync/doc-definitions.js (hunk 3 above)
Comment (dkichler) [inline databases/business-sync/doc-definitions.js]: Indeed, under the `channels` property:  https://github.com/Kashoo/synctos#essential-document-properties

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 4 above)
Comment (dkichler) [inline databases/business-sync/fragment-shoebox-snippet.js]: Not a bad idea, I'll look into the flexibility of changes to enums.  However, this field should probably also enforce that the value matches the type segment from the keyspace.

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 5 above)
Comment (dkichler) [inline databases/business-sync/fragment-shoebox-snippet.js]: I think a `received` field would be useful.  

On the topic of data updates - I'm not entirely sure how we should handle this scenario, perhaps a history array for deltas and their received timestamp.  Perhaps we can leave this as a future improvement for now.

Comment (dkichler) [inline databases/business-sync/fragment-shoebox-snippet.js]: As for processed, see my comment above at creating a new record when we process one of these raw documents.

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 1 above)
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: Nice thinking, I like it! Any processing data would be great in a different document, for example a `banking-reconciliation` document... 👍

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 2 above)
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: My suggestions are to go with either shoeboxItem or sourceData. I prefer sourceData on a 6/10 strength :D I'd rather keep the metaphor out of the key names because it could become dated quickly and might need to be explained to every new dev. On the other hand, if we're always talking shoebox, then shoeboxItem/Snippet might come to mind first.

Does @cclark have any thoughts?

File: databases/business-sync/doc-definitions.js (hunk 3 above)
Comment (ziemek) [inline databases/business-sync/doc-definitions.js]: 👍

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 2 above)
Comment (dkichler) [inline databases/business-sync/fragment-shoebox-snippet.js]: Personally, I like the metaphor as it fits with our 'story' - maybe not a bad thing to have to explain to every new dev.  Personal preference, really, but I could get behind `shoeboxItem`.

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 4 above)
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: 👍

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 5 above)
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: `received` timestamp for sure. Would help us track when the data came into our system, which can be correlated with outages or the source code at the time of the import. 👍

Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: I'm flip-flopping on a `processed` boolean. Without it, it wouldnt be trivial to tell whether an item was already read by the system or not, it would take a query (`select item where id not in processed-items...`). This may also be inefficient (for example, for bank rec, `select all items where item.id not in processed).

On the other hand, it may be inefficient to update each shoebox item whenever it's actually processed (`update item set processed = true`...). This could be done as a bulk call to couchbase but it would still touch a bunch of random bytes throughout the table...

Any more thoughts in reply to mine?

Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: For updates, it crossed my mind that we could possibly treat them as separate shoebox items and then just correlate them in the same processed document. For example, a changed square item could would have two shoebox items, but they would be processed into the same kashoo record. thoughts?

Can definitely leave it as a future improvement though 👍

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 4 above)
Comment (dkichler) [inline databases/business-sync/fragment-shoebox-snippet.js]: Looks like those two features are mutually exclusive.  We can have an enum, but can't then also enforce that the field value is the same as the keyspace segment.  Unless we extend synctos a little more, at least.

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 5 above)
Comment (dkichler) [inline databases/business-sync/fragment-shoebox-snippet.js]: Yeah, I can't say I fully on board with a `processed` flag.  You make some good points, but is a simple flag sufficient?  Will there only be a single time we process any given item?  It might be good to track artifacts of processing, but maybe be more flexible as an array of associated processed document IDs (and maybe the processed timestamp as well?  although that would certainly also be a field in the 'processed' document)

Comment (dkichler) [inline databases/business-sync/fragment-shoebox-snippet.js]: I'm now also wondering if an identifier from the external source should be another field

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 4 above)
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: Probably more important to ensure that the type property matches what's in the key. Although, what's the use case for having the type property in the document and not just in the key? Curious...

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 5 above)
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: +1 on adding a source identifier.

Agreed that we can leave out `processed` for now and add it in the future to improve any queries if we actually find that it is warranted.

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 4 above)
Comment (cclark) [inline databases/business-sync/fragment-shoebox-snippet.js]: Anytime I come across a `type` field it feels like it scream enum to me so I agree with Ziemek's hunch. I know with synctos it will be easier to have the type safety check. I think the question comes down to how challenging they are to evolve and refactor. Probably not a lot different than in Java where it is doable but not something you undertake lightly (although the implied backing by a sequence of integers probably doesn't exist and makes it easier).

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 1 above)
Comment (cclark) [inline databases/business-sync/fragment-shoebox-snippet.js]: As I was working through the types with Amrik I came to the exact same conclusion. This really is the junk drawer. Things get thrown in here and we don't know much about them so I'm not sure it is worthwhile trying to store info about the Kashoo entity it gets connected to.

As it gets processed down the pipeline the next document (`banking-reconciliation`, `tax-filing`) might reference this document as the original source but will have all of the information about the "processing" section of the pipeline with the metadata that Dave mentions. I think this document is for the initial portion of the pipeline, "collection" into the shoebox.

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 4 above)
Comment (dkichler) [inline databases/business-sync/fragment-shoebox-snippet.js]: Personally I like to have the field available on the entity rather than needed to deduce from the key (at least when programming against the entity).  I think it also gives more options for querying (ie, a N1QL query with an index on that field may be faster than applying a regex across all keys.  Or maybe not, but at least we have both options to weigh against each other)

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 5 above)
Comment (cclark) [inline databases/business-sync/fragment-shoebox-snippet.js]: Definitely a 👍 on a timestamp. Do we have an agreed convention (e.g. `created`, `createdAt`, `creationTime`)? In this instance I think there is a creation/received timestamp in our system that is definitely different than the transaction time (when the transaction was booked in Square).

Updates are a good question. I think I'm tending to defer on this one for now although my gut says there might be an array of updates, each of which contains just the fields of the document that have deltas and the new value. That is probably a pre-mature optimization as opposed to having a whole document. And each would also have a receivedAt (or agreed name from previous section) time to know when the update was received.

We are going to need it. Square has confirmed that already. But let's take a KISS/YAGNI attitude and come back to it.

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 4 above)
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: good reasoning 👍

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 5 above)
Comment (cclark) [inline databases/business-sync/fragment-shoebox-snippet.js]: In my list of potential properties I had wondered if there was an "encoding" that might need to be stored. This might be the `Content-Type` of the response when we're loading a txn via a webhook notification. It might be XML, it might be JSON, maybe it is some crazy binary format. Perhaps @ziemek 's investigation into Yodlee RESTful URLs and JSON vs XML will inform this. I could be convinced it is added complexity we do not need until we get our first item which is not JSON.

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 1 above)
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: I'm on board with `shoeboxItem`. `sourceData` is definitely too abstract

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 2 above)
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: I'm on board with `shoeboxItem`. `sourceData` is definitely too abstract 👍

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 5 above)
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: I like the content-type suggestion though i'm not yet sure how it would apply to something like emails and receipts. I'd vote deferring it until we need it and have it implicitly be json for now.

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 4 above)
Comment (cclark) [inline databases/business-sync/fragment-shoebox-snippet.js]: Agreed on having it available on the entity.

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 2 above)
Comment (cclark) [inline databases/business-sync/fragment-shoebox-snippet.js]: Decision made. We need the :gavel: at GitHub

File: databases/business-sync/fragment-shoebox-snippet.js (hunk 5 above)
Comment (cclark) [inline databases/business-sync/fragment-shoebox-snippet.js]: Definitely a +1 on a source identifier.

I can't wrap my head around `processed` yet. It does seem like we will want to have something which will allow us to easily return the ones that need to be reviewed and processed. If it is automation that is doing it then efficiency of the query isn't necessarily a concern at this point. However, if we're looking to show a shoebox listing of items that need attention we'll be looking to pull the unprocessed items regularly and quickly.

Another twist to throw into this, and it may apply more to the documents representing that processed view. Let's say the ML determines that CostCo is normally "Office Supplies", printer paper and snacks for the office. It can see that your average CostCo bill is $300. But then you have a CostCo transaction that is $5,000. You purchased a laptop, phone and printer for the business. This shoebox item may have been processed to be categorized as "Office Supplies", however, because the total is so different than other transactions we flag the processing as "Draft/Needs Confirmation" and explain that because of the value of the transaction we think it may need closer attention. Perhaps the business owner classifies it as "Capital Asset" or something else to show it needs to be tracked and depreciated. Maybe they don't care and they leave it as "Office Supplies". After the user confirms then it really is processed.

This feels like something that we'll address later as a part of processing and perhaps just gives fuel to the argument that `processed` is deferred for now but I wanted to bring up the scenario in case it enables someone else to bring clarity to what we're modeling.

Comment (cclark) [inline databases/business-sync/fragment-shoebox-snippet.js]: Agreed. Defer for now. I'm not sure if emails show HTML vs. plaintext. And for binary objects like a picture of a receipt, I'm assuming we'd pass that off to the blobstore and have a reference to it here instead of encoding it uuencoded or similar here.

Comment (ziemek) [inline databases/business-sync/fragment-shoebox-snippet.js]: In your example, I think the details would not be surfaced here. I imagined that the ML algorithm would pull all `bank` type records that do not have processed = true. In this example, it would create its own intermediate record with internal state and would mark the shoeboxItem as processed (because it already has relevant processing data, whatever it is).

Of course the fact that the doc is processed is also implied by the processing data document existing, so this goes back to this boolean being either used to optimize queries or perhaps for programmer ease-of-use when looking at the docs.

**Comment by dkichler** [comment]
Integrated all suggestions/feedback so far, with the exception of changing the `type` field to an enum - this can be done in a future commit or even PR when we have a list of the first basic types we'd like to support.  For now it remains an unconstrained string.

File: databases/business-sync/fragment-shoebox-item.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      required: true,
+      mustNotBeEmpty: true
+    },
+    source: {
+      // The source of the item
+      type: 'string',
+      required: true,
+      mustNotBeEmpty: true
+    },
+    sourceId: {
```
Comment (ziemek) [inline databases/business-sync/fragment-shoebox-item.js]: What is the use case for having the source id surfaced here? I'm not sure it will be helpful, would like to chat it through.

Comment (dkichler) [reply databases/business-sync/fragment-shoebox-item.js]: Basically to support tracing it back.  A redundancy measure in the rare case that a received entity does not contain the ID as a field (CB documents are like this).  You had previously given your support to such a field:
> +1 on adding a source identifier.

**Comment by ziemek** [comment]
Nice one! :shipit:
